mmeval/metrics/perplexity.py

```diff
@@ -0,0 +1,281 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+# This class is modified from `torchmetrics
+# <https://github.com/Lightning-AI/metrics/blob/master/src/torchmetrics/text/perplexity.py>`_.
+import numpy as np
+from typing import TYPE_CHECKING, Dict, List, Sequence, Tuple, Union, overload
+
+from mmeval import BaseMetric
+from mmeval.core.dispatcher import dispatch
+from mmeval.utils import try_import
+
+if TYPE_CHECKING:
+    import oneflow
+    import oneflow as flow
+    import paddle
+    import tensorflow
+    import tensorflow as tf
+    import torch
+else:
+    paddle = try_import('paddle')
+    torch = try_import('torch')
+    tf = try_import('tensorflow')
+    flow = try_import('oneflow')
+
+
+def softmax(x):
+    """This function is used to calculate softmax for np and jnp."""
+    exps = np.exp(x - np.max(x, axis=1, keepdims=True))
+    return exps / np.sum(exps, axis=1, keepdims=True)
+
+
+class Perplexity(BaseMetric):
+    """Perplexity measures how well a language model predicts a text sample.
+
+    It is commonly used as a metric for evaluating the quality of a
+    language model.It is defined as 2 to the power of the cross-entropy
+    loss of the model(or the negative log-likelihood of the sample).
+
+    Args:
+        ignore_labels (Union[int, List[int], None]): Integer specifying a
+            target class to ignore. If given, this class index does not
+            contribute to the returned score. Defaults to None.
+        **kwargs: Keyword parameters passed to :class:`BaseMetric`.
+
+    Examples:
+        >>> from mmeval import Perplexity
+        >>> predictions = np.asarray([1,1,1,1])
+        >>> targets = np.asarray([1,1,1,1])
+        >>> metric = Perplexity()
+        >>> result = metric(predictions, targets)
+        {'perplexity': 0.5226045319355426}
+    """
+
+    def __init__(self,
+                 ignore_labels: Union[int, List[int], None] = None,
+                 **kwargs) -> None:
+        super().__init__(**kwargs)
+        self.ignore_labels = None
+        if isinstance(ignore_labels, int):
+            ignore_labels = [ignore_labels]
+        if ignore_labels is not None:
+            for ignore_label in ignore_labels:
+                if not isinstance(ignore_label, int):
+                    raise ValueError(
+                        f'Argument `ignore_labels` expected to be `None`, `int`,`List[int]` but got {ignore_labels}'  # noqa: E501
+                    )
+            self.ignore_labels = list(set(ignore_labels))
```

Comment:
We can use the `mmeval.utils.is_list_of`:
```suggestion
        if isinstance(ignore_labels, int):
            ignore_labels = [ignore_labels]
        if ignore_labels is not None:
            if not is_list_of(ignore_labels, int):
                raise ValueError(
                    f'Argument `ignore_labels` expected to be `None`, `int`,`List[int]` but got {ignore_labels}'  # noqa: E501
                )
            ignore_labels = list(set(ignore_labels))
        self.ignore_labels = ignore_labels
```